packages/nx/src/plugins/js/project-graph/build-dependencies/target-project-locator.ts

```diff
@@ -1,45 +1,74 @@
-import {
-  getRootTsConfigFileName,
-  resolveModuleByImport,
-} from '../../utils/typescript';
-import { isRelativePath, readJsonFile } from '../../../../utils/fileutils';
-import { dirname, join, posix, relative, resolve } from 'path';
-import { workspaceRoot } from '../../../../utils/workspace-root';
+import { readFileSync } from 'node:fs';
+import { builtinModules } from 'node:module';
+import { dirname, join, posix, relative } from 'node:path';
 import {
   ProjectGraphExternalNode,
   ProjectGraphProjectNode,
 } from '../../../../config/project-graph';
-import { builtinModules } from 'module';
 import {
   createProjectRootMappings,
   findProjectForPath,
 } from '../../../../project-graph/utils/find-project-for-path';
+import { isRelativePath, readJsonFile } from '../../../../utils/fileutils';
+import { parseJson } from '../../../../utils/json';
+import { workspaceRoot } from '../../../../utils/workspace-root';
+import { findExternalPackageJsonPath } from '../../utils/find-external-package-json-path';
+import {
+  getRootTsConfigFileName,
+  resolveModuleByImport,
+} from '../../utils/typescript';
 
-const builtInModuleSet = new Set<string>([
+/**
+ * The key is a combination of the package name and the project root importing it
+ * e.g. `lodash__packages/my-lib`, the value is the resolved external node name
+ * from the project graph.
+ */
+export type NpmResolutionCache = Map<string, string>;
+
+export const builtInModuleSet = new Set<string>([
   ...builtinModules,
   ...builtinModules.map((x) => `node:${x}`),
 ]);
 
+export function isBuiltinModuleImport(importExpr: string): boolean {
+  const packageName = parsePackageNameFromImportExpression(importExpr);
+  return builtInModuleSet.has(packageName);
+}
+
 export class TargetProjectLocator {
   private projectRootMappings = createProjectRootMappings(this.nodes);
-  private npmProjects = filterRootExternalDependencies(this.externalNodes);
+  private npmProjects = Object.keys(this.externalNodes)
+    .filter((k) => k.startsWith('npm:'))
+    .map((k) => this.externalNodes[k]);
   private tsConfig = this.getRootTsConfig();
   private paths = this.tsConfig.config?.compilerOptions?.paths;
   private typescriptResolutionCache = new Map<string, string | null>();
-  private npmResolutionCache = new Map<string, string | null>();
 
   constructor(
     private readonly nodes: Record<string, ProjectGraphProjectNode>,
-    private readonly externalNodes: Record<string, ProjectGraphExternalNode>
+    private readonly externalNodes: Record<
+      string,
+      ProjectGraphExternalNode
+    > = {},
+    private readonly npmResolutionCache: NpmResolutionCache = new Map<
```

Comment:
I previously responded to the module level cache here: https://github.com/nrwl/nx/pull/23307#discussion_r1604868361

- I didn't want to run any risks of stale caches via require caching
- Unit tests are more straightforward/direct because the cache can easily be stubbed/mocked

Reply:
After updating the code to apply the other changes such as sharing a single targetProjectLocator instance, I combined these ideas by applying a default cache at the module level which is still overridable via the constructor parameter for unit tests